### PR TITLE
PMM-7380 backup location list add

### PR DESCRIPTION
**Pull request by Dasio**
[PMM-7380](https://jira.percona.com/browse/PMM-7380)

- [ ] Build: https://github.com/Percona-Lab/pmm-submodules/pull/1458

- [ ] Links to other linked pull requests (optional).
- API: https://github.com/percona/pmm/pull/674
- API tests: https://github.com/Percona-Lab/pmm-api-tests/pull/181
---
- [ ] Tests passed.
- [ ] Feature build pass.
- [ ] (Re)requested review.
- [ ] Fix conflicts with target branch.
- [ ] Update API dependency.
- [ ] Update jira ticket description if needed.
- [ ] Attach screenshots/console output to confirm new behavior to jira ticket, if applicable.

### CODE REVIEW

File: models/location_helpers_test.go [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+				},
+			},
+			errorMsg: "",
+		},
+		{
+			name: "fs config - missing path",
+			location: models.CreateBackupLocationParams{
+				Name: "fs-1",
+				FSConfig: &models.FSLocationConfig{
+					Path: "rpc error: code = InvalidArgument desc = FS path field is empty",
```
Comment (oter) [inline models/location_helpers_test.go:137]: Maybe it should be in `errorMsg`?

File: models/location_helpers_test.go [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+		loc1, err := models.CreateBackupLocation(q, params1)
+		require.NoError(t, err)
+		loc2, err := models.CreateBackupLocation(q, params2)
+		require.NoError(t, err)
+
+		actual, err := models.FindBackupLocations(q)
+		require.NoError(t, err)
+		var found1, found2 bool
+		for _, channel := range actual {
```
Comment (BupycHuk) [inline models/location_helpers_test.go:96]: channel?

File: models/location_helpers_test.go [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+		require.NoError(t, sqlDB.Close())
+	}()
+	db := reform.NewDB(sqlDB, postgresql.Dialect, reform.NewPrintfLogger(t.Logf))
+
+	tests := []struct {
+		name     string
+		location models.CreateBackupLocationParams
+		errorMsg string
+	}{
+		{
```
Comment (BupycHuk) [inline models/location_helpers_test.go:122]: can we add two more tests:
1. with both configs are passed
2. no config is passed

File: models/location_helpers_test.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+)
+
+func TestBackupLocations(t *testing.T) {
+	sqlDB := testdb.Open(t, models.SkipFixtures, nil)
+	defer func() {
+		require.NoError(t, sqlDB.Close())
+	}()
+	db := reform.NewDB(sqlDB, postgresql.Dialect, reform.NewPrintfLogger(t.Logf))
+
+	t.Run("create", func(t *testing.T) {
```
Comment (BupycHuk) [inline models/location_helpers_test.go:38]: can we add the same test for S3?

File: models/location_helpers_test.go [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+		q := tx.Querier
+
+		params1 := models.CreateBackupLocationParams{
+			Name:        "some name",
+			Description: "some desc",
+			FSConfig: &models.FSLocationConfig{
+				Path: "/tmp",
+			},
+		}
+		params2 := models.CreateBackupLocationParams{
```
Comment (BupycHuk) [inline models/location_helpers_test.go:80]: let's create two different types here

File: services/management/backup/locations_service.go [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+	backupv1beta1 "github.com/percona/pmm/api/managementpb/backup"
+	"github.com/pkg/errors"
+	"gopkg.in/reform.v1"
+
+	"github.com/percona/pmm-managed/models"
+)
+
+// LocationsService represents backup locations API.
+type LocationsService struct {
```
Comment (BupycHuk) [inline services/management/backup/locations_service.go:30]: Can we add some tests to this service?

**Review by BupycHuk — CHANGES_REQUESTED** [review]
Please use `PMM-XXXX` format in PR title

File: models/location_helpers.go [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+// CreateBackupLocation persists backup location.
+func CreateBackupLocation(q *reform.Querier, params CreateBackupLocationParams) (*BackupLocation, error) {
+	id := "/location_id/" + uuid.New().String()
+
+	if err := checkUniqueLocationID(q, id); err != nil {
+		return nil, err
+	}
+
+	if params.Name == "" {
```
Comment (BupycHuk) [inline models/location_helpers.go:105]: this check can be done on api and database labels, do we really need it here too?

File: models/location_helpers_test.go [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+			},
+			S3Config: &models.S3LocationConfig{
+				Endpoint:  "https://example.com/bucket",
+				AccessKey: "access_key",
+				SecretKey: "secret_key",
+			},
+		}
+
+		_, err = models.CreateBackupLocation(q, params)
+		require.NotNil(t, err)
```
Comment (BupycHuk) [inline models/location_helpers_test.go:118]: let's use `require.EqualError`

File: models/location_helpers_test.go [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+		loc1, err := models.CreateBackupLocation(q, params1)
+		require.NoError(t, err)
+		loc2, err := models.CreateBackupLocation(q, params2)
+		require.NoError(t, err)
+
+		actual, err := models.FindBackupLocations(q)
+		require.NoError(t, err)
+		var found1, found2 bool
+		for _, location := range actual {
+			if location.ID == loc1.ID {
```
Comment (BupycHuk) [inline models/location_helpers_test.go:156]: let's check other fields too.
Also you can use `assert.Condition`

File: services/management/backup/locations_service_test.go [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+		},
+	})
+	require.Nil(t, err)
+
+	t.Run("list", func(t *testing.T) {
+		res, err := svc.ListLocations(ctx, &backupv1beta1.ListLocationsRequest{})
+		assert.Nil(t, err)
+
+		assert.Len(t, res.Locations, 2)
+	})
```
Comment (BupycHuk) [inline services/management/backup/locations_service_test.go:116]: let's check field values

**Review by BupycHuk — CHANGES_REQUESTED** [review]

File: models/location_helpers.go (hunk 7 above)
Comment (Dasio) [reply models/location_helpers.go:105]: I copied it from other same existing functionality. I'm removing it then.

File: main.go [hunk 11]
```diff
@@ -42,8 +42,10 @@ import (
 	"github.com/percona/pmm/api/agentpb"
 	"github.com/percona/pmm/api/inventorypb"
 	"github.com/percona/pmm/api/managementpb"
+	backupv1beta1 "github.com/percona/pmm/api/managementpb/backup"
 	dbaasv1beta1 "github.com/percona/pmm/api/managementpb/dbaas"
 	iav1beta1 "github.com/percona/pmm/api/managementpb/ia"
+
```
Comment (artemgavrilov) [inline main.go:48]: Redundant line

File: models/location_helpers.go [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"gopkg.in/reform.v1"
+)
+
+func checkUniqueLocationID(q *reform.Querier, id string) error {
+	if id == "" {
+		panic("empty Channel ID")
```
Comment (artemgavrilov) [inline models/location_helpers.go:29]: Location ID

File: models/location_helpers.go [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+		return errors.WithStack(err)
+	}
+}
+
+func checkFSConfig(c *FSLocationConfig) error {
+	if c == nil {
+		return status.Error(codes.InvalidArgument, "FS location config is empty.")
+	}
+	if c.Path == "" {
+		return status.Error(codes.InvalidArgument, "FS path field is empty")
```
Comment (artemgavrilov) [inline models/location_helpers.go:48]: Missing dot at the end of the error message.

File: models/location_helpers_test.go [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+			},
+			S3Config: &models.S3LocationConfig{
+				Endpoint:  "https://example.com/bucket",
+				AccessKey: "access_key",
+				SecretKey: "secret_key",
+			},
+		}
+
+		_, err = models.CreateBackupLocation(q, params)
+		require.EqualError(t, err, "rpc error: code = InvalidArgument desc = Only one config is allowed")
```
Comment (artemgavrilov) [inline models/location_helpers_test.go:118]: tests.AssertGRPCError

File: services/management/backup/locations_service_test.go [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+
+	svc := NewLocationsService(db)
+	t.Run("add fs", func(t *testing.T) {
+		loc, err := svc.AddLocation(ctx, &backupv1beta1.AddLocationRequest{
+			Name: gofakeit.Name(),
+			FsConfig: &backupv1beta1.FSConfig{
+				Path: "/tmp",
+			},
+		})
+		assert.Nil(t, err)
```
Comment (artemgavrilov) [inline services/management/backup/locations_service_test.go:51]: assert.NoError

**Review by artemgavrilov — COMMENTED** [review]
Minor issues

File: services/management/backup/locations_service_test.go (hunk 10 above)
Comment (artemgavrilov) [reply services/management/backup/locations_service_test.go:151]: I agree with Nurlan

**Review by artemgavrilov — APPROVED** [review]
In general LGTM

File: services/management/backup/locations_service_test.go (hunk 10 above)
Comment (Dasio) [reply services/management/backup/locations_service_test.go:151]: It's already there.

File: models/location_helpers_test.go (hunk 14 above)
Comment (BupycHuk) [reply models/location_helpers_test.go]: Even better

**Review by BupycHuk — APPROVED** [review]

**Review by oter — APPROVED** [review]

**Review by artemgavrilov — APPROVED** [review]

**Pull request merged**